[ZIP 312] Specify key generation, change randomizer handling

conradoplg opened the pull request:
Changes to randomizer handling after I realized that it couldn't possibly be generated using the message as an input. It does simplify things too.

Some additional fixes (using "Re-Randomized FROST" consistently; updating Owners)

fixes #1051

conradoplg committed:
ZIP 312: change randomizer handling

str4d commented on zips/zip-0312.rst:
```diff
@@ -212,11 +175,11 @@ involves these steps:
-authenticated channel.)
+In Round Two, the Coordinator picks a message and sends it to each signer along
+with the commitments received in Round One, over an authenticated and encrypted
+channel. In Re-Randomized FROST, an additional value called `randomizer` needs
+to be generated by the Coordinator and sent along with the message and
+commitments.
```
Does the math not work out if each participant generates a share of alpha and commits to it in round 1, then uses their share in round 2? If it didn't then I guess the way to remove the trust assumption on the Coordinator here would be to still do that commitment in round 1, and then have a round 1.5 that is just participants revealing their alpha shares, but that would turn this into a three-round protocol, or two-round with preprocessing (where participants precompute and save their shares of both R and alpha), which isn't ideal for some usage scenarios where the Coordinator _is_ trusted. Maybe that should be left as an option?

daira replied:
One way to reduce the requirements on the Coordinator is to have it generate the randomizer by hashing the message, commitments, and an extra seed. The participants then recalculate it using that hash. The Coordinator can grind on the seed in order to create a subliminal channel, but this means you're not trusting the Coordinator's RNG. (Also, no participant can grind on their contribution in order to influence the randomizer, because the Coordinator only reveals the seed after it has the message and commitments.)

conradoplg replied:
I tried @daira 's approach in 62833e6c

str4d commented on zips/zip-0312.rst:
```diff
@@ -228,6 +191,14 @@ that the given SIGHASH matches the data sent from the Coordinator, or compute th
 SIGHASH themselves from that data. However, the specific mechanism for that process
 is outside the scope of this ZIP.
```
I think the mechanism for using FROST for transaction spend auth should _not_ be outside the scope of this ZIP, given that this ZIP is "FROST for Spend Authorization Multisignatures" (rather than the smaller-scoped "Re-randomized FROST"). However, I agree that most of the content that mechanism would entail is what would go into creating and checking PCZTs, so what we should do here is incorporate by reference the ZIP that eventually contains that specification (https://github.com/zcash/zips/issues/693), and then either use it as appropriate directly in this section, or divide the specification into "Re-randomized FROST" (for more general consumption) and "Using Re-randomized FROST for Zcash transactions" (which then augments it further with the Zcash-specific logic like PCZT usage).

conradoplg replied:
I agree, let's leave this pending until we get more concrete on PCZTs

daira replied:
Oh it's even worse. That comment is *only visible on the "new experience" view*.

daira commented on rendered/zip-0312.html:
```diff
@@ -10,11 +10,11 @@
 Status: Draft
 Category: Wallet
-Created: 2022-08-dd
+Created: 2024-08-dd
```
This can just be the date it was originally committed which was apparently 2023-01-12. https://github.com/zcash/zips/commit/48f8c2c80e890385a9619e4dbf3b62ff8963b010
```suggestion
Created: 2023-01-12
```

daira commented on zips/zip-0312.rst:
```diff
@@ -3,11 +3,11 @@
   Status: Draft
   Category: Wallet
-  Created: 2022-08-dd
+  Created: 2024-08-dd
```
This can just be the date it was originally committed which was apparently 2023-01-12. https://github.com/zcash/zips/commit/48f8c2c80e890385a9619e4dbf3b62ff8963b010
```suggestion
Created: 2023-01-12
```

daira commented on zips/zip-0312.rst:
```diff
@@ -86,7 +86,7 @@ shielded transaction:
   as an auxiliary (secret) input, among others.
 
-When employing re-randomizable FROST as specified in this ZIP, the goal is to
+When employing re-randomized FROST as specified in this ZIP, the goal is to
```
Right, because we're describing the variant that *is* re-randomized.

daira commented on zips/zip-0312.rst:
```diff
@@ -153,50 +153,13 @@ using seed phrases to recover the original secret (which may be something
-
 Round One - Commitment
 ''''''''''''''''''''''
 
```
Typo just below here: "Roune One" -> "Round One"

conradoplg replied:
thanks, solved in 62833e6c

daira commented on zips/zip-0312.rst:
```diff
@@ -228,6 +191,14 @@ that the given SIGHASH matches the data sent from the Coordinator, or compute th
+generated accordingly to the Zcash specification [#protocol-spendauthsig]_. Note
+that the choice of :math:`\mathsf{alpha}` influences the SIGHASH computation, so
+it is impossible to compute the `randomizer` based on the message (SIGHASH), as
+suggested in [#frost-rerandomized]_. This is not an issue as long the
+`randomizer` is generated uniformly at random, which is done in the Zcash
+specification.
```
It's an overspecification to imply that the randomizer needs to be generated uniformly at random. As suggested above, it could be chosen depending on the commitments sent in round 1.

Also, the spec actually says to generate the randomizer like this:

> Define $\mathsf{RedDSA.GenRandom} ⦂ () \rightarrow \mathsf{RedDSA.Random}$ as:
> * Choose a byte sequence $T$ uniformly at random on $\mathbb{B}^{\mathbb{Y}[(\ell_{\mathsf{H}} + 128) / 8]}$.
> * Return $\mathsf{H}^⊛(T)$.

To refine my suggestion, $T$ would be the "seed" and then the randomizer would be computed by all participants as $\mathsf{H}^⊛(T \hspace{0.2em}\mathsf{||}\hspace{0.2em} \mathsf{commitments}).$

conradoplg replied:
Great idea. I went ahead with it in 62833e6c

daira reviewed (dismissed):
ACK with suggestions.

conradoplg committed:
Update zips/zip-0312.rst

Co-authored-by: Daira-Emma Hopwood <daira@jacaranda.org>

conradoplg committed:
update randomizer generator to be derived from signing commitments

str4d commented on zips/zip-0312.rst:
```diff
@@ -163,44 +159,68 @@ following modifications.
-Implementations MAY choose another encoding as long as all values (the message,
-and the identifier, binding nonce and hiding nonce for each participant) are
+Two new helper functions are defined, which generate a randomizer; the first is
+used by the Coordinator and the second is used by participants. The
+`encode_signing_commitments` function SHOULD return the byte serialization of
```
This is somewhat confusing, given that `random_bytes` gets its own paragraph. Separate `encode_signing_commitments` into its own paragraph, and then maybe turn those two paragraphs into bullet points? or something that clarifies these are _references_ to functions defined elsewhere that are used by the two new helper functions, not the two new helper functions being defined. Maybe also import the references first, and then say "Two new helper functions are defined" immediately before the actual definitions.

conradoplg replied:
Agreed, I restructured this in c74bed50

str4d commented on zips/zip-0312.rst:
```diff
@@ -163,44 +159,68 @@ following modifications.
 Randomizer Generation
 '''''''''''''''''''''
```
Move this above `Re-Randomized FROST` and make it a same-level section. So this ZIP's Specification section then has:
- Key Generation
- Randomizer Generation
- Re-Randomized FROST
  - Round One - Commitment
  - Round Two - Signature Share Generation
  - Signature Share Verification and Aggregation
- Ciphersuites
  - FROST(Jubjub, BLAKE2b-512)
  - FROST(Pallas, BLAKE2b-512)

which then means that the "Re-Randomized FROST" sub-section is just focused on modifications to FROST, and everything else is stuff canonically specified in ZIP 312 (although maybe parts of "Key Generation" move elsewhere to e.g. ZIP 32, but we'll figure that out in #883).

conradoplg replied:
Good point, done in c74bed50

str4d commented on zips/zip-0312.rst:
```diff
@@ -163,44 +159,68 @@ following modifications.
+`encode_signing_commitments` function SHOULD return the byte serialization of
+the `commitment_list` value as the serialization of the map between identifiers
+and signing commitments, as described in [#frost-serialization]_.
+Implementations MAY choose another encoding as long as all values (the
+identifier, binding nonce and hiding nonce for each participant) are
 unambiguously encoded.
```
There is a risk here that a Coordinator and Participant (or multiple Participants) might have different implementations, and attempt to use them in the same signing session. That would be incompatible and cause a failure, which violates the defintion of MAY in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt):
> 5. MAY   This word, or the adjective "OPTIONAL", mean that an item is
   truly optional.  One vendor may choose to include the item because a
   particular marketplace requires it or because the vendor feels that
   it enhances the product while another vendor may omit the same item.
   An implementation which does not include a particular option MUST be
   prepared to interoperate with another implementation which does
   include the option, though perhaps with reduced functionality. In the
   same vein an implementation which does include a particular option
   MUST be prepared to interoperate with another implementation which
   does not include the option (except, of course, for the feature the
   option provides.)

So we will need to specify the full encoding here.

conradoplg replied:
Done in c74bed50 by using the serialization function from the RFC

str4d commented on zips/zip-0312.rst:
```diff
@@ -422,6 +443,22 @@ signing generates a re-randomized signature:
 
+Regarding randomizer handling, in Zcash, the `randomizer` is called
+:math:`\mathsf{alpha}` and is usually generated using the `RedDSA.GenRandom`
+function as defined inthe  Zcash specification [#protocol-spendauthsig]_. Note
```
```suggestion
function as defined in the Zcash specification [#protocol-spendauthsig]_. Note
```

conradoplg committed:
Update zips/zip-0312.rst

Co-authored-by: Jack Grigg <jack@electriccoin.co>

conradoplg committed:
address comments; use serialization function from the RFC

conradoplg reviewed (commented):
Addressed comments in did some additional improvements/cleanups

nuttycom commented:
[force-pushed](https://github.com/zcash/zips/compare/42b086e3d6547e6335f3b5026a7481d2ad23cf03..0d08f3f00f5162cf8258cd54831e49fd2ee4e68b#diff-4030bb7d56803845f7f30f22cfd915248371e96a351aa1cd7ba71d0ca6fc3e39) to resolve conflicts with `main`.

daira commented on zips/zip-0312.rst:
```diff
@@ -153,33 +149,54 @@ using seed phrases to recover the original secret (which may be something
+- The `Ns` constant is define in [#frost-primeordergroup]_ and is the size of a
+  serialized scalar.
+- The `H2(m)` function is a ciphersuite-generic function defined in
+  [#frost-hash]_ but it is instantiated in the Ciphersuites section.
```
```suggestion
  [#frost-hash]_ but it is instantiated in the `Ciphersuites`_ section.
```

daira commented on zips/zip-0312.rst:
```suggestion
The key word "MUST" in this document is to be interpreted as described
in BCP 14 [#BCP14]_ when, and only when, it appears in all capitals.
```

daira commented on zips/zip-0312.rst:
```diff
@@ -153,33 +149,54 @@ using seed phrases to recover the original secret (which may be something
-Implementations MAY choose another encoding as long as all values (the message,
-and the identifier, binding nonce and hiding nonce for each participant) are
-unambiguously encoded.
+  Outputs: (randomizer_seed, randomizer), a byte buffer and a Scalar
```
```suggestion
  Outputs: (randomizer_seed, randomizer), a byte sequence and a Scalar
```

daira commented on zips/zip-0312.rst:
```diff
@@ -153,33 +149,54 @@ using seed phrases to recover the original secret (which may be something
-with `n` bytes sampled uniformly at random. The constant `Ns` is also specified
-in [#FROST]_ and is the size of a serialized scalar.
+  def randomizer_generate(commitment_list):
+    # Generate a random byte buffer with the size of a serialized scalar
```
```suggestion
    # Generate a random byte sequence with the size of a serialized scalar.
```

daira commented on zips/zip-0312.rst:
```diff
@@ -189,18 +206,23 @@ in [#FROST]_ and is the size of a serialized scalar.
+-------------------
+
+To add re-randomization to FROST, follow the specification [#FROST]_ with the
+following modifications.
```
```suggestion
To add re-randomization to FROST, the signers MUST follow the specification
[#FROST]_ with the following modifications.
```

daira commented on zips/zip-0312.rst:
```diff
@@ -436,14 +462,17 @@ References
 .. [#frost-serialization] `The ZF FROST Book, Serialization Format <https://frost.zfnd.org/user/serialization.html>`_
+.. [#frost-listoperations] `RFC 9591: The Flexible Round-Optimized Schnorr Threshold (FROST) Protocol for Two-Round Schnorr Signatures. Section 4.3: List Operations <https://www.rfc-editor.org/rfc/rfc9591.html#name-list-operations>`_
+.. [#frost-conventions] `RFC 9591: The Flexible Round-Optimized Schnorr Threshold (FROST) Protocol for Two-Round Schnorr Signatures. Section 2: Conventions and Definitions <https://www.rfc-editor.org/rfc/rfc9591.html#name-conventions-and-definitions>`_
+.. [#frost-hash] `RFC 9591: The Flexible Round-Optimized Schnorr Threshold (FROST) Protocol for Two-Round Schnorr Signatures. Section 3.2: Cryptographic Hash Function <https://www.rfc-editor.org/rfc/rfc9591.html#name-cryptographic-hash-function>`_
```
```suggestion
.. [#frost-hash] `RFC 9591: The Flexible Round-Optimized Schnorr Threshold (FROST) Protocol for Two-Round Schnorr Signatures. Section 3.2: Cryptographic Hash Function <https://www.rfc-editor.org/rfc/rfc9591.html#name-cryptographic-hash-function>`_
.. [#frost-sign] `RFC 9591: The Flexible Round-Optimized Schnorr Threshold (FROST) Protocol for Two-Round Schnorr Signatures. Section 5.2: Round Two - Signature Share Generation <https://www.rfc-editor.org/rfc/rfc9591.html#name-round-two-signature-share-g>`_
```

daira commented on zips/zip-0312.rst:
```diff
@@ -228,9 +251,9 @@ that the given SIGHASH matches the data sent from the Coordinator, or compute th
+`randomizer_regenerate()`).
 
 - ``sk_i = sk_i + randomizer``
 - ``group_public_key = group_public_key + G.ScalarBaseMult(randomizer)``
```
```suggestion
Let `FROST.sign` be the regular FROST signing function with arguments
`(identifier, sk_i, group_public_key, nonce_i, msg, commitment_list)`,
as defined in [#frost-sign]_. The `randomized_sign` function takes an
additional input `randomizer_seed`, and adjusts the inputs relative to
the `randomizer` computed by `randomizer_regenerate()`:

::

  def randomized_sign(identifier, sk_i, group_public_key, nonce_i, msg,
                      commitment_list, randomizer_seed):
    randomizer = randomizer_regenerate(randomizer_seed, commitment_list)
    randomized_sk_i = sk_i + randomizer
    randomized_group_public_key = group_public_key + G.ScalarBaseMult(randomizer)
    return FROST.sign(identifier, randomized_sk_i, randomized_group_public_key,
                      nonce_i, msg, commitment_list)
```

daira replied:
This should use double-backticks. (I'd fix it but my browser is broken.)

daira commented on zips/zip-0312.rst:
```suggestion
Let ``FROST.verify_signature_share`` be the regular FROST signature share
verification function with arguments
``(identifier, PK_i, comm_i, sig_share_i, commitment_list, group_public_key, msg)``,
as defined in [#frost-aggregate]_. The ``randomized_verify_signature_share``
function takes an additional input ``randomizer`` computed as above, and adjusts
the inputs relative to this randomizer:

::

  def randomized_verify_signature_share(identifier, PK_i, comm_i, sig_share_i,
                                        commitment_list, group_public_key, msg):
    group_offset = G.ScalarBaseMult(randomizer)
    randomized_PK_i = PK_i + group_offset
    randomized_group_public_key = group_public_key + group_offset
    return FROST.verify_signature_share(identifier, randomized_PK_i, comm_i, sig_share_i,
                                        commitment_list, randomized_group_public_key, msg)
```

daira committed:
Apply suggestions from @daira's review.

Signed-off-by: Daira-Emma Hopwood <daira@jacaranda.org>

daira committed:
ZIP 312: Add @daira as an Owner.

Signed-off-by: Daira-Emma Hopwood <daira@jacaranda.org>

daira committed:
ZIP 312: Add this PR to the Pull-Request header.

ZIP 0: Document that the Pull-Request header uses the singular spelling
even if there are multiple PRs (there are several existing cases of this).

Signed-off-by: Daira-Emma Hopwood <daira@jacaranda.org>

daira committed:
ZIP 312: Cosmetics.

Signed-off-by: Daira-Emma Hopwood <daira@jacaranda.org>

daira committed:
ZIP 312: Expand on the disadvantages of trusted dealer key generation.
(In my opinion it completely defeats the point of a threshold signature
scheme.)

Signed-off-by: Daira-Emma Hopwood <daira@jacaranda.org>

daira committed:
ZIP 312: Modifications for Quantum Recoverability.
(This might be moved into another PR.)

Signed-off-by: Daira-Emma Hopwood <daira@jacaranda.org>

daira committed:
ZIP 312: Adapt @str4d's changes from #883. Also reorder, correct, and
update references.

Signed-off-by: Daira-Emma Hopwood <daira@jacaranda.org>

nuttycom commented on zips/zip-0312.rst:
```diff
@@ -126,60 +131,165 @@ Specification
+from a common spending key $\mathsf{sk}$ as described in
+[#protocol-orchardkeycomponents]_ [#protocol-saplingkeycomponents]_, with the
+exception of $\mathsf{ask}$ which SHOULD NOT be derived from $\mathsf{sk}$ (and
+can't possibly be derived, if Distributed Key Generation was used).
```
Remove the redundancy between here and the paragraph starting at line 201.

nuttycom commented on zips/zip-0312.rst:
```diff
@@ -126,60 +131,165 @@ Specification
+$\mathsf{ak}$, $\mathsf{nk}$, and $\mathsf{rivk\_ext}$.
+
+Since Sapling will not be modified to support Quantum Recoverability
+[#draft-ecc-quantum-recoverability]_, Orchard is RECOMMENDED for new FROST keys.
```
Note that `#draft-ecc-quantum-recoverability` provides the rationale for this choice.

nuttycom commented on zips/zip-0312.rst:
```diff
@@ -126,60 +131,165 @@ Specification
+participants; their entire hardware and software stack also needs to be trusted
+by all participants. Since this will often be an unrealistic assumption that
+partly defeats the point of using a threshold signature scheme, trusted dealer
+key generation SHOULD NOT be used.
```
```suggestion
Trusted dealer key generation has the severe disadvantage that it
requires absolutely trustworthy hardware and software to be used by the dealer
in order to reliably avoid exposing the private key material. That is, it is not
sufficient simply to have an individual or organization that is trusted by all
participants; their entire hardware and software stack also needs to be trusted
by all participants. Since this will often be an unrealistic assumption that
partly defeats the point of using a threshold signature scheme, trusted dealer
key generation SHOULD NOT be used.
```

And make this a WARNING section

daira committed:
zip-guide[-markdown]: Correct the syntax for notes and warnings.

Signed-off-by: Daira-Emma Hopwood <daira@jacaranda.org>

daira committed:
ZIP 312: Update the Abstract to be more precise about re-randomizability
as a prerequisite for unlinkability of Sapling and Orchard spend
authorization signatures.

Signed-off-by: Daira-Emma Hopwood <daira@jacaranda.org>

daira commented on zips/zip-0312.rst:
```diff
@@ -126,60 +131,162 @@ Specification
+  can be sent along with the FROST shares during key generation. This option MAY
+  be used if Distributed Key Generation was used, if participants find it
+  acceptable to trust one participant to correctly generate $\mathsf{sk}$.
+- Participants generate a $\mathsf{sk}$ in distributed manner using TODO.
```
This should also reference https://c2sp.org/cocktail-dkg .

str4d reviewed (commented):
Flushing my review comment from a while back on 9c24f216572c44eb6ca6f509f9f7ae242331d92d.

daira committed:
ZIP 312: Wording improvements related to Quantum Recoverability.

Signed-off-by: Daira-Emma Hopwood <daira@jacaranda.org>

daira committed:
ZIP 312: Minor wording updates.

Signed-off-by: Daira-Emma Hopwood <daira@jacaranda.org>

daira committed:
ZIP 312: Use COCKTAIL-DKG.

Signed-off-by: Daira-Emma Hopwood <daira@jacaranda.org>

daira committed:
ZIP 312: Cosmetics.

Signed-off-by: Daira-Emma Hopwood <daira@jacaranda.org>

str4d commented on zips/zip-0312.rst:
```diff
@@ -3,24 +3,29 @@
 
-The terms below are to be interpreted as follows:
+The term "Recovery Protocol" in this document is to be interpreted as described
+in [#draft-ecc-quantum-recoverability]_.
```
Update this to point to ZIP 2005 instead.

str4d commented on zips/zip-0312.rst:
```diff
@@ -101,21 +109,27 @@ With those considerations in mind, the threat model considered in this ZIP is:
+
+Caveats concerning the security model for use of FROST keys in a Quantum
+Recovery protocol are covered in
+[#draft-ecc-quantum-recoverability-usage-with-frost]_.
```
Update this reference.

str4d commented on zips/zip-0312.rst:
```diff
@@ -126,131 +140,367 @@ Specification
-multiplication, where the protocol spec would use $[k] P$ with the group
-implied by $P$.
-
-An additional per-ciphersuite hash function is used, denote ``HR(m)``, which
```
Remove the corresponding instantiations of `HR` below.

daira committed:
ZIP 312: Update references from draft-ecc-quantum-recoverability to ZIP 2005.

Signed-off-by: Daira-Emma Hopwood <daira@jacaranda.org>

str4d commented on zips/zip-0312.rst:
```diff
@@ -126,131 +139,366 @@ Specification
+participants do not provide different commitment values to a subset of
+participants), over a secure broadcast channel [#frost-term-broadcast-channel]_.
+
+Other Distributed Key Generation protocols MAY be used.
```
@conradoplg Does the ZF reference impl actually implement ChillDKG (on which COCKTAIL-DKG is based), or does it implement the DKG from the original FROST paper? If the latter, is that actually compatible with ChillDKG?

Without confirmation to the affirmative (especially because COCKTAIL-DKG is not finished), we should not claim it:
```suggestion
A suitable DKG protocol is COCKTAIL-DKG [#cocktail-dkg]_.
COCKTAIL-DKG assumes that each participant has a static public key known to all
other participants.

Other Distributed Key Generation protocols MAY be used. For example, the Zcash
Foundation's reference implementation of FROST [#zf-frost-dkg]_ includes a DKG
implementation for some ciphersuites. As noted in its documentation, the reference
implementation requires participants to broadcast the commitment values honestly
(i.e. participants do not provide different commitment values to a subset of
participants), over a secure broadcast channel [#frost-term-broadcast-channel]_.
```

EDIT: https://github.com/ZcashFoundation/frost/blob/main/frost-core/src/keys/dkg.rs makes it clear that it implements the DKG from the original FROST paper. So we need confirmation about whether or not that is exactly compatible with COCKTAIL-DKG.

conradoplg replied:
Apologies, I missed that. No we don't support COCKTAIL-DKG (though you could build it on top of the libraries)

str4d commented on zips/zip-0312.rst:
```diff
@@ -126,131 +139,366 @@ Specification
+
+A contribution list is a value of the form $[(i, \mathtt{contribution}_{\,i}),\, \ldots]$.
+Each element in the list indicates a NonZeroScalar identifier $i$ and a byte
+sequence $\mathtt{contribution}_{\,i}$. This list MUST be sorted in ascending
```
Clarify that all `contribution_i`s in a contribution list are the same length (so the encoding is well-defined).

conradoplg replied:
Now moot

str4d commented on zips/zip-0312.rst:
```diff
@@ -126,131 +139,366 @@ Specification
+The parties agree on $\mathsf{sk}$ as follows:
+
+1. The Coordinator chooses a random seed
+   $\mathtt{sk\_seed} = \mathtt{random\_bytes}(\mathtt{Nh})$.
```
Bug: this `sk_seed` is not used.
```suggestion
```

str4d commented on zips/zip-0312.rst:
```diff
@@ -126,131 +139,366 @@ Specification
+   encrypted, authenticated channel.
+4. All parties regenerate $\mathsf{sk}$ using
+   $\mathtt{sk\_regenerate}(\mathtt{sk\_seed}, \mathtt{contribution\_list})$,
+   and they include a Pedersen commitment to it in the transcript.
```
Bug: the other participants cannot run this, as they do not learn the `contribution_i` values from the other participants due to (in the current spec) them not being broadcast.

conradoplg replied:
Done in be04583, though we need to discuss the exact mechanism

str4d reviewed (dismissed):
Flushing comments as of afc247b90f1b289b89bffd6585e3acef1deaa1c0 from ZIP Sync with @daira @nuttycom @arya2.

conradoplg committed:
address comments and small cleanups

conradoplg committed:
fill section on contributory generation of sk

conradoplg commented on zips/zip-0312.rst:
```diff
@@ -126,131 +139,366 @@ Specification
+- Participants generate a $\mathsf{sk}$ in a contributory manner, for example
+  using the method defined in `Contributory Generation of sk`_. This approach
+  SHOULD be used to generate $\mathsf{sk}$ if Distributed Key Generation was
+  employed.
```
Done in 31661fb00

conradoplg commented:
> The next step is to rebase out [f9e077d](https://github.com/zcash/zips/commit/f9e077d54dca485c6cdae832f32f0b180a24cdaf) and [afc247b](https://github.com/zcash/zips/commit/afc247b90f1b289b89bffd6585e3acef1deaa1c0) since those changes cannot work, and instead do what @str4d says in [#895 (comment)](https://github.com/zcash/zips/pull/895#discussion_r2625036304) . I cannot do that on my phone; can someone else do it?

I (finally) did this and addressed @str4d comments

ValarDragon commented on zips/zip-0312.rst:
```diff
@@ -126,60 +139,161 @@ Specification
+When generating FROST key components for the Orchard shielded protocol, the
+remaining parts of the Orchard key tree apart from $\mathsf{ask}$ and
+$\mathsf{ak}$ SHOULD be generated as described in [#zip-2005-usage-with-frost]_.
+The specification there is normative, and is only summarized here. It requires
+the participants to privately agree on a value $\mathsf{sk}$ (which is no longer
+sufficient by itself for spend authorization), and then use it to obtain the
+nullifier deriving key $\mathsf{nk}$, the quantum spending key $\mathsf{qsk}$,
```
Isn't `sk` sufficient for spend authorization in the QR recovery case though? 

So in the recovery protocol, the threshold security assumption drops?

ValarDragon replied:
Perhaps its `sk` + `quantum-computer access`, so with a QC, any one multisig member could steal funds.

If this understanding is correct, we should explicitly note this. I wonder if its worth suggesting to use generic MPC tools like SPDZ w/ PQ-OT. (I think this should not block merge of the ZIP, and just be a note. We can ask folks like stoffelMPC)

ValarDragon replied:
cref: https://x.com/badcryptobitch/status/2055777683343130849?s=20

daira replied:
ZIP 2005 (https://zips.z.cash/zip-2005#usagewithfrost) already says that.

> Isn't sk sufficient for spend authorization in the QR recovery case though?

No, because you also need to sign for the FROST group key. If you have a CRQC then you can do that (but that is precisely what ZIP 2005 says).

nuttycom commented:
It looks like version 0.2 of https://c2sp.org/cocktail-dkg has now landed, so anything that that was blocking here can now be reassessed and the ZIP can be updated.

str4d commented on zips/zip-0312.rst:
```diff
@@ -189,68 +303,127 @@ in [#FROST]_ and is the size of a serialized scalar.
+- In Round 1, generate ``sk_i = random_bytes(32)``
+- Also in Round 1, set $\mathsf{payload}_{i,j} = \mathtt{sk\_i}$ for all `j`
+- When running Round 3, use the following as the ``extension`` argument:
+  $ext = H(n \parallel len(payload_1) \parallel payload_1 \parallel \cdots \parallel len(payload_n) \parallel payload_n)$
```
Check whether this was meant to refer to a COCKTAIL-DKG payload (i.e. committing to other parts of the transcript) or `sk_i` (i.e. just the randomness from each participant). If the latter, then each `sk_i` could potentially be HD-derived instead of random, although there is already additional information that each signer needs to store beside their wallet seed phrase in order to use a threshold account.

str4d replied:
From a call today with @conradoplg and nikkolasg: If we do a pure hash of concatenated `sk_i` then the last participant can bias `sk` (though to actually benefit from it, they would need to grind to bias `nk` or `qsk` via their chained hashes, as `sk` isn't used for anything directly). A way to avoid this issue could be to take the COCKTAIL-DKG transcript as the final input (like the random beacon we used in the Sapling proving key DKG).

Relatedly, nikkolasg had asked why not just derive `sk` from `ak`. I pointed out that this would break Sapling proof authorization keys, and it would mean that Orchard FVKs are fully derivable solely from `ak` instead of the FVK components being independent. (In current practice, `rivk` is derived from the other components due to ZIP 2005, but (`ak, nk, qk)` are still independent.)

Given that `ak` is sort-of-but-not-quite derived from the transcript, it might be fine to use `ak` as the final "random beacon" input instead of (a digest of) the transcript. It would still mean that `(nk, qk)` are in part derived from `ak`, but the fresh randomness from `sk_i` might mean they are sufficiently independent? I'm gonna think about it.
